src/app/modules/attachment/attachment.service.ts

```diff
@@ -0,0 +1,51 @@
+import crypto from 'crypto'
+import { ResultAsync } from 'neverthrow'
+
+import { aws as AwsConfig } from '../../../config/config'
+import { createLoggerWithLabel } from '../../../config/logger'
+import { AttachmentUploadError } from '../core/core.errors'
+
+const logger = createLoggerWithLabel(module)
+
+export const uploadAttachments = (
+  formId: string,
+  attachmentData: Record<string, unknown>,
+): ResultAsync<Map<string, string>, AttachmentUploadError> => {
+  const attachmentMetadata = new Map()
+  const attachmentUploadPromises = []
+
+  for (const fieldId in attachmentData) {
+    const individualAttachment = JSON.stringify(attachmentData[fieldId])
+
+    const hashStr = crypto
+      .createHash('sha256')
+      .update(individualAttachment)
+      .digest('hex')
+
+    const uploadKey =
+      formId + '/' + crypto.randomBytes(20).toString('hex') + '/' + hashStr
+
+    attachmentMetadata.set(fieldId, uploadKey)
+    attachmentUploadPromises.push(
+      AwsConfig.s3
+        .upload({
+          Bucket: AwsConfig.attachmentS3Bucket,
+          Key: uploadKey,
+          Body: Buffer.from(individualAttachment),
+        })
+        .promise(),
+    )
+  }
+
+  return ResultAsync.fromPromise(
+    Promise.all(attachmentUploadPromises).then(() => attachmentMetadata),
+    (error) => {
+      logger.error({
+        message: 'S3 attachment upload error',
+        meta: { action: 'uploadAttachments' },
+        error,
+      })
+      return new AttachmentUploadError()
+    },
+  )
```

Comment:
just as a matter of convention, could you use `.map(() => attachmentMetadata)` here instead of the `Promise.then` above? this makes it slightly easier for other developers to find what is being returned, since we use `.map` for this use case in most of the rest of the codebase